### ISSUE TITLE
logged in authentication 

### ISSUE THREAD

**Issue by josesanches1** [Azure-Samples/ms-identity-java-servlet-webapp-authentication#3]
When logged into Azure AD, the validateState method in the AuthHelper class throws an exception "indicates state param mismatch, null, empty or expired".
How it can be resolved.
thanks

**Comment by idg-sam** [comment]
Hi @josesanches1 ,

1. Thanks for the bug report. Can you please give some more detail as to when this error is triggered, and in which one of the samples in this repository?
2. Please note that a state mismatch exception is expected in the following scenarios:
    1. When the application (.war file) is very recently re-deployed, it causes all saved session data (including the saved `state` param) to be removed. Therefore, the freshly-deployed app might not have a record of  the `state` param that AAD sends back into the `auth/redirect` endpoint. If your use-case requires a work-around, it will be dependent on your specific set up. By default, Tomcat stores session data in-memory and does not survive re-deployment. You may modify your solution so that your session variables are stored in a non-volatile manner, such as a separate cache, database, or file system. For example, see [Tomcat session replication](https://tomcat.apache.org/tomcat-9.0-doc/cluster-howto.html).
    1. When the sign-in has been initiated from a link to AAD's `/authorize` endpoint that was **not generated from our application**. Therefore the app won't have a record of  the `state` param that AAD sends back into the `auth/redirect` endpoint. This feature mitigates against CSRF attacks.
    1.  When sign-in has been initiated from a link to AAD's `/authorize` endpoint that was **generated from our application but already used to complete a sign-in**. Saved state values are cleared if there is a successful sign-in or some sign-in errors; therefore, the app won't have a record of  the `state` param that AAD sends back into the `auth/redirect` endpoint.
    1. When the user has completed their  sign-in flow from a different browser than the one from which they initiated the sign-in, or cleared cookies, or changed their session ID in between. The app ties the saved `state` param to a browser session; therefore, the app won't have a record of  the `state` param that AAD sends back into the `auth/redirect` endpoint.
    1. When the user starts another sign-in flow before completing the first, from the same browser session, and then tries to complete the first one (e.g., the user has multiple open of the app and has triggered sign-in from both of them). The app ties only one saved `state` param to a browser session; therefore, the app won't have a record of  the first sign-in flow's `state` param that AAD sends back into the `auth/redirect` endpoint (only the latest one is persisted). If this scenario is important to you, you may choose to save a list of valid state params rather than only the one from the most recently-triggered sign-in flow (see the `state` property in the `IdentityContextData` class and the `validateState` method in the `AuthHelper` class).
    1. When the user has taken more than 10 minutes between initiating and completing sign-in. In this case, the the app considers the `state` param that AAD sends back into the `auth/redirect` endpoint is *expired*. You may adjust this time limit from the authentication.properties file.
    
Regards,
@idg-sam

**Comment by glammers1** [comment]
Hi! Taking advantage of this ongoing discussion, I would like to ask a question here. I saw a comment from @josesanches1 in another sample [repository](https://github.com/Azure-Samples/ms-identity-java-webapp/issues/70#issuecomment-791906451) that uses Spring Security along with Azure AD, saying that the best option is to use the approach followed here (`ms-identity-java-servlet-webapp-authentication`) and forget about Spring.

I have some reservations regarding that affirmation, though. In my humble opinion, one of the best options is to use the library `azure-spring-boot-starter-active-directory`. You can find an example [here](https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp).

However, for those who cannot use that library because of not having the latest versions of Spring Framework, I wonder whether the solution proposed in this repository could be used in production or it is just for learning purposes. Let's say you have to decide about:

Having Spring Framework 3.0.x (you already have it and cannot upgrade it for now), then you would need to add Spring Security 3.0.x + Spring Security Oauth2 (a version from 2012 more or less to meet your Spring libraries version constraints).

**vs**

Create a solution like the one showed in this repository where the developer does the whole implementation based on this sample project using `MSAL4J`.

TL;DR Is this repo just a sample focus on showing how `MSAL4J` works or is it also an alternative for those who don't want or cannot use a Spring-based solution? I really like to know your opinion @idg-sam since I've seen you as a contributor in several different sample projects for Azure AD (and some of them use Spring).

Thanks. I'm sorry in advance if you don't expect a question like this here.

**Comment by idg-sam** [comment]
Great question @glammers1 

- If I was starting a web app from scratch, I would use the combo of SpringMVC+SpringSecurity+ActiveDirectory boot starters as [demonstrated in this document](https://docs.microsoft.com/azure/developer/java/spring-framework/configure-spring-boot-starter-java-app-with-azure-active-directory). This option will abstracts away much of the details of signing in, obtaining tokens, protecting endpoints with, and implement of some of the more advanced scenarios with MSAL4J (e.g., [On-Behalf-Of flow](https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-resource-server-obo)). In short, by utilizing this boot starters combination, you'll get the advantages of Spring Security framework and of MSAL4J. See [this Spring Blog post](https://spring.io/blog/2021/01/13/the-latest-on-azure-active-directory-integration) for all available features.

- If you're tied to an older version of Spring Security and can't take advantage of the Active Directory boot starter, you're still covered by Spring Security and Spring OAuth2 for the most common use-cases except for scenarios like the On-Behalf-Of flow. If you need OBO, you can implement it fairly simply (see [here for an example of Spring protected resource server + manual implementation of OBO flow with MSAL4J](https://github.com/Azure-Samples/ms-identity-java-webapi)). In this case, the Java Servlet samples in this repository won't provide much benefit.

- The purpose of the samples in this repository (Java Servlets) is to showcase  how to benefit from Azure Active Directory and Microsoft Identity Platform features and functionality in Java web applications where using Spring Security for authentication/authorization isn't an option. These samples showcase how to keep track of user authentication status, extract and make use of claims, protect routes by authentication status, protect routes by roles and security group claims (all of which are trivial when using Spring Security), how to inject Access Tokens to Graph requests, and how to deploy your app to Azure app service.

TL;DR: For building web applications and protected web APIs, you'd likely want to:
1. Use ActiveDirectory+SpringSecurity5 boot starters if possible (we're in the process of making a chapter-wise tutorial for this).
2. Use SpringSecurity & Spring OAuth2.0 if the previous option is not available to you.
3. Use manual implementation of MSAL4J as demonstrated in this repo as a guide for your app, if the previous two options are not available to you.

PS - We are currently working on a chapter-wise tutorial format for showcasing samples utilizing the Azure Active Directory Spring Boot Starter as well. These will guide the developer to learn how to utilize the starter, beginning from the simplest use-cases and ramping up the more advanced scenarios incrementally.

**Comment by glammers1** [comment]
Hi @idg-sam thanks for the detailed answer, I really appreciated it.

I agree with you on your ranking, and I think that points 1 and 3 are clear for someone who chooses any of them. That's why I want to focus on point 2 a little bit more.

As you said it covers the most use-cases except for scenarios that the one you mentioned. I can think of another scenario which is the [Groups overage claim](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens). Points 1 and 3 are already prepared out-of-the-box to deal with that scenario whereas in point 2 I think you would wind up using somewhat `MSAL4J` to overcome it. Apart from that issue, it also seems that the developer has to tinker with `SpringSecurity` config to limit access to endpoints based on `Azure Groups`. For instance, testing point 1 from this [sample](https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-webapp) I was able to use groups for authorization because it is already integrated into the library. Taking a look at the authorities array inside the authentication object, which is an instance of `Oauth2AuthenticationToken` in this case, I see my groups there:

<img width="547" alt="Screenshot 2021-03-16 at 11 50 18" src="https://user-images.githubusercontent.com/13570903/111304581-d17a7780-864d-11eb-934a-3721c80958eb.png">

Nevertheless, using any other sample that doesn't use SpringSecurity5 boot starters, groups are not properly mapped in the authorities array. The next screenshot comes from this [sample](https://github.com/Azure-Samples/ms-identity-java-webapp/tree/master/spring-security-web-app), where the authentication object is an instance of `Oauth2Authentication`:

<img width="529" alt="Screenshot 2021-03-16 at 11 55 37" src="https://user-images.githubusercontent.com/13570903/111305197-890f8980-864e-11eb-80d6-bbdfea3706dc.png">

As you can see, we don't have `group1` there even though it comes in the token.

So I would say you would have to create token filters and manage user detail objects by yourself somehow. I saw this [message](https://github.com/Azure-Samples/ms-identity-java-webapi/pull/17#issuecomment-702962550) from you talking about this some months ago. Do you happen to know if a fork was created to address security group based access control for those cases? I'm asking because if the code the developer has to write, plus dealing with MSAL4J if needed, increases the complexity of approach number 2, maybe option 3 becomes option 2.

Thanks!

**Comment by idg-sam** [comment]
Hi @glammers1 , thanks again for another great question.

Up until this point we planned to create chapter-wise tutorials covering non-spring framework use-cases with pure MSAL4J and Java Servlets (this servlets tutorial repo), and another set of chapter-wise tutorials for spring 5 (with the active directory boot starter), and retire the older spring samples. However, you bring up a good point that we need to address legacy Spring users as well. I'll bring this up at my next meeting with my PMs today.

I will give this use-case (using groups and other ID token claims in legacy Spring apps) some thought and tinker with some code to see what might be a good solution.

In the interim, *when using legacy spring security + spring OAuth*  and using security group membership and other ID token claims with Azure AD, consider either of: 
- the solution as seen in chapter 3.2 of this tutorial (copy the `AuthHelper` and `IdentityContextData` classes and implement a custom `IdentityContextAdapter` interface if needed)
- look into [this solution](https://www.baeldung.com/spring-security-openid-connect-legacy) for inspiration.

Please note that I haven't yet tried either of these and can't definitively prefer either option over the other until I implement and test the options thoroughly.

Regards,
@idg-sam

**Comment by josesanches1** [comment]
Hello,
Thank you for the explanation.
The bug was detected in the "1-Authentication sing-in" project.
The state UUID value is not saved in the session.
When the comparison is made in validateState the value of IdentityContextData is empty.
IdentityContextData should be in session to perform the comparison. Whenever an IdentityContextData parameter is updated, the session variable should be updated.

In the tests I did, implementing the update of IdentityContextData in the session variable (whenever there is a change in the parameters) solved this problem.

Yes, I am aware that when the application server is restarted, the session value is reset to null. But, the application server has not been restarted.
In the tests I did not reopen the browser, the coockies are the same.
I understand the implementation made validateNonce.
Thanks

**Comment by josesanches1** [comment]
Hello glammers1,

I like to have maximum control over the code's actions.
But I respect different opinions.
As for this example, I think it is quite good to be adapted in production.

**Comment by glammers1** [comment]
@josesanches1 I see your point. This sample is awesome. It helped me understand how things work behind the scenes. And as you said, you have all the control over the whole flow if needed.

**Comment by idg-sam** [comment]
Hi @josesanches1 ,

> In the tests I did, implementing the update of IdentityContextData in the session variable (whenever there is a change in the parameters) solved this problem.
> 

Thanks for investigating this. Can you please elaborate about your fix? I was under the impression that [this event listener](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-authentication/blob/main/1-Authentication/sign-in/src/main/java/com/microsoft/azuresamples/msal4j/helpers/IdentityContextAdapterServlet.java#L45) in IdentityContextAdapter for Servlets will take care of saving the session data automatically (and it seems to be behaving as expected on my set up).

1. When you were experiencing the bug, were you using this same adapter implementation that I linked? 
2. Which version of JDK are you using?
3. Which version of Tomcat are you using?

**Comment by josesanches1** [comment]
Hi Sam

My environment:
tomcat 8.5.61 x86, JDK 1.8.213 x86 and Eclipse 2018 x86

The bug I identified is related to the update of the session.
when the setContext method of the IdentityContextAdapter servlet is invoked, the state parameter is null.
The getParameter method of the IdentityContextAdapterServlet servlet is fine and with the correct value.
Given this situation, I forced the update of IdentityContextData in the session, when the value is assigned to the state parameter, it resolved.
thanks

**Comment by idg-sam** [comment]
I will make a draft PR to manually save context after [this line](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-authentication/blob/62765c25fb69ff2ab49ee8f6817a15c07c813882/1-Authentication/sign-in/src/main/java/com/microsoft/azuresamples/msal4j/helpers/AuthHelper.java#L108) in each of the samples.

In our testing this hasn't been a problem so far but I have been using a different JDK version and Tomcat version.

**Comment by josesanches1** [comment]
Hi @idg-sam 
I appreciate all the attention you have shown on this topic.
Keep up the great work.
thank you so much

**Comment by kilokili777** [comment]
> Hi @josesanches1 ,
> 
> I will make a draft PR to manually save context after [this line](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-authentication/blob/62765c25fb69ff2ab49ee8f6817a15c07c813882/1-Authentication/sign-in/src/main/java/com/microsoft/azuresamples/msal4j/helpers/AuthHelper.java#L108) in each of the samples. I think that might resolve the issue you're facing?
> 
> In our testing this hasn't been a problem so far but I have been using a different JDK version and Tomcat version.
> 
> Regards, @idg-sam

What happened to the draft PR?
